dawgz/schedulers.py

```diff
@@ -241,20 +221,32 @@ async def _submit(self, job: Job) -> str:
             lines.append('#')
 
         ## Dependencies
-        separator = '?' if job.waitfor == 'any' else ','
-        keywords = {
-            'success': 'afterok',
-            'failure': 'afternotok',
-            'any': 'afterany',
-        }
+        jobids = await asyncio.gather(*[
+            self.submit(dep)
+            for dep in job.dependencies
+        ])
 
-        deps = [
-            f'{keywords[status]}:{jobid}'
-            for jobid, (_, status) in zip(jobids, job.dependencies.items())
-        ]
+        deps = []
+
+        for jobid, (dep, status) in zip(jobids, job.dependencies.items()):
+            if jobid is None:  # dep is done
+                if status == 'failure':
+                    if job.waitfor == 'all':
+                        raise DependencyNeverSatisfiedError(f'aborting {job}') from JobNotFailedError(f'{dep}')
+                elif job.waitfor == 'any':
+                    deps.clear()
+                    break
+            elif isinstance(jobid, Exception):
+                raise DependencyNeverSatisfiedError(f'aborting {job}') from jobid
+            else:
+                deps.append(f'{self.deptype[status]}:{jobid}')
+        else:
+            if job.waitfor == 'any' and not deps:
+                raise DependencyNeverSatisfiedError(f'aborting {job}')
```

Comment:
This logic needs review.

Reply:
Can you provide a minimal example where Slurm submits such an empty job (on 0.1.6)? Then I can verify this solution and the fix I pushed in #3

Reply:
```python
@ensure(lambda: True)
@job
def empty():
    pass

@after(empty, status='failure')
@job
def notok():
    pass
```

In `master` and `0.1.6`, we cannot prune `empty` from the dependencies of `notok` because it waits for failure. Then, it will create an empty slurm job. Instead, what I propose is to make `submit` instantly return `None` when a job is done. With that I am able to check within the scheduler whether there is a `DependencyNeverSatisfiedError`.

Reply:
Tbh, I don't think moving (partial) pruning responsibility to the scheduler is the approach we should be taking. The scheduler should only submit the requested graph. In fact, the example you've provided has the desired behavior in #3 and alters the pruning procedure alone.